oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
+                                                                          : () -> issueAccessToken(token);
+            } else {
+                tokenIssuingFunc = () -> this.tokenFuture;
+            }
+
+            final CompletionStage<GrantedOAuth2AccessToken> tokenIssuingFuture = RequestContext.mapCurrent(
+                            ctx -> ctx.makeContextAware(tokenIssuingFunc.get()), tokenIssuingFunc);
+            tokenIssuingFuture.handle((newToken, cause) -> {
+                if (cause != null) {
+                    future.completeExceptionally(cause);
+                } else {
+                    future.complete(newToken);
+                    if (saveTokenFunc != null) {
+                        // suppress 'ReturnValueIgnored' warning
+                        final CompletableFuture<Void> unused = saveTokenFunc.apply(newToken);
+                    }
+                }
+                return null;
+            });
+            return null;
+        });
+        return future;
+    }
+
+    private boolean isValidToken(@Nullable GrantedOAuth2AccessToken token) {
+        return token != null && token.isValid(Instant.now().plus(refreshBefore));
+    }
+
+    private CompletableFuture<GrantedOAuth2AccessToken> issueAccessToken(
+            @Nullable GrantedOAuth2AccessToken token) {
+        if (token == null && loadTokenFunc != null) {
+            return loadTokenFunc.get().thenCompose(storedToken -> {
+                if (storedToken.isValid()) {
+                    return UnmodifiableFuture.completedFuture(storedToken);
+                }
+                return obtainAccessToken(null);
+            });
+        }
+        return obtainAccessToken(token);
+    }
+
+    private CompletableFuture<GrantedOAuth2AccessToken> refreshAccessToken(GrantedOAuth2AccessToken token) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        refreshRequest.make(token).handle((newToken, cause) -> {
+            if (cause != null) {
+                if (cause instanceof TokenRequestException) {
+                    // try to issue a new access token from scratch
+                    final CompletableFuture<GrantedOAuth2AccessToken> tokenUpdateFuture =
+                            RequestContext.mapCurrent(ctx -> ctx.makeContextAware(obtainAccessToken(token)),
+                                                      () -> issueAccessToken(token));
+                    tokenUpdateFuture.handle((newToken0, cause0) -> {
+                        if (cause0 != null) {
+                            future.completeExceptionally(cause0);
+                        } else {
+                            future.complete(newToken);
+                        }
+                        return null;
+                    });
+                } else {
+                    future.completeExceptionally(cause);
+                }
```

Comment:
Would it be better to extract exception handling logic using `.exceptionally(...)`?